---
title: "IBS_microbiome_V6_WGCNA"
author: "Swapna Mahurkar-Joshi"
date: "October 7, 2016"
output: word_document
---
```{r}
library(WGCNA)
library(phyloseq)
library(FunctionsAndLibrariesSJ)
library(cluster)
```

Significant OTU from deseq can be clustered and tested for association with traits using WGCNA
```{r}
##
# microbiome matrix
load("C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSmicrobiomeProject/Resequencing/batch1_batch2_merged/miRNA_microbiome_integration11232016/microbMat.Rda")

# microbiome mapping
littleIbsMappingIBsHcClust<-read.delim(file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSmicrobiomeProject/Resequencing/batch1_batch2_merged/New_qiime_analyses_11062016/littleIbsMappingIBsHcClust.csv",sep=",", row.names=1); dim(littleIbsMappingIBsHcClust)
row.names(littleIbsMappingIBsHcClust)[1:4] <- c("Biopsy.A6379" , "Biopsy.A6244",  "Biopsy.A5169",  "Biopsy.A5501" )


# setwd("C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSmicrobiomeProject/Resequencing/batch1_batch2_merged/phyloseq_workflow") 
# sigMuc <- read.csv("sig_adjP_Group.csv", sep = ",", row.names = 1)
# load("PS2")
# dat1 <- otu_table(PS2)[row.names(otu_table(PS2))%in%row.names(sigMuc),]
# trait1 <- sample_data(PS2)
setwd("C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSmicrobiomeProject/Resequencing/batch1_batch2_merged/phyloseq_workflow/WGCNA_microbiome") 

##
# colnames(microbMat) <- paste(matrix(unlist(strsplit(colnames(microbMat), ";")), nrow=dim(microbMat)[2], ncol= 7, byrow = TRUE)[,2], matrix(unlist(strsplit(colnames(microbMat), ";")), nrow=dim(microbMat)[2], ncol= 7, byrow = TRUE)[,6], sep = ";")

colnames(microbMat) <- paste(matrix(unlist(strsplit(colnames(microbMat), ";")), nrow=dim(microbMat)[2], ncol= 7, byrow = TRUE)[,2], matrix(unlist(strsplit(colnames(microbMat), ";")), nrow=dim(microbMat)[2], ncol= 7, byrow = TRUE)[,6], sep = ";")

A=adjacency(microbMat, type="distance")
hierADJ1 = hclust(as.dist(1-A),method="complete")

## windows(width=8, height=4)
png("heirarchical_microb.png", bg="white")
# par(mar=c(1,2.5,5,1))
plot(hierADJ1,  xlab="", sub="");
dev.off()
dat1 <- t(microbMat)
trait1 <- littleIbsMappingIBsHcClust
row.names(trait1) <- substr(row.names(trait1),8,12)
match(row.names(trait1), colnames(dat1))
```

# IBS and HC separately
```{r}
dat1.ibs <- dat1[,colnames(dat1)%in%row.names(subset(trait1, trait1$Group == 2))]; dim(dat1.ibs)
dat1.hc <- dat1[,colnames(dat1)%in%row.names(subset(trait1, trait1$Group == 1))]; dim(dat1.hc)
nSets = 2
multiExpr = vector(mode = "list", length = nSets)
multiExpr[[1]] = list(data = as.data.frame(t(dat1.ibs)));
# names(multiExpr[[1]]$data) = femData$substanceBXH;
multiExpr[[2]] = list(data = as.data.frame(t(dat1.hc)));

# Check that the data has the correct format for many functions operating on multiple sets:
exprSize = checkSets(multiExpr)

# Check that all genes and samples have sufficiently low numbers of missing values.
gsg = goodSamplesGenesMS(multiExpr, verbose = 3);
gsg$allOK

if (!gsg$allOK)
{
  # Print information about the removed genes:
  if (sum(!gsg$goodGenes) > 0)
    printFlush(paste("Removing genes:", paste(names(multiExpr[[1]]$data)[!gsg$goodGenes], 
                                              collapse = ", ")))
  for (set in 1:exprSize$nSets)
  {
    if (sum(!gsg$goodSamples[[set]]))
      printFlush(paste("In set", setLabels[set], "removing samples",
                       paste(rownames(multiExpr[[set]]$data)[!gsg$goodSamples[[set]]], collapse = ", ")))
    # Remove the offending genes and samples
    multiExpr[[set]]$data = multiExpr[[set]]$data[gsg$goodSamples[[set]], gsg$goodGenes];
  }
  # Update exprSize
  exprSize = checkSets(multiExpr)
}

sampleTrees = list()
for (set in 1:nSets)
{
  sampleTrees[[set]] = hclust(dist(multiExpr[[set]]$data), method = "average")
}
setLabels <- c("IBS","HCs")
png(file = "SampleClustering.png", width = 2000, height = 3000, res = 350);
par(mfrow=c(2,1))
par(mar = c(0, 4, 2, 0))
for (set in 1:nSets)
  plot(sampleTrees[[set]], main = paste("Sample clustering on all OTUs in", setLabels[set]),
       xlab="", sub="", cex = 0.7);
dev.off();

```

```{r}
# Choose the "base" cut height for the IBS data set
baseHeight = 0.358
# Adjust the cut height for the hc data set for the number of samples
cutHeights = c(0.358, 0.358*exprSize$nSamples[2]/exprSize$nSamples[1]);
# Re-plot the dendrograms including the cut lines
png(file = "SampleClustering_0.37cutoff.png", width = 2000, height = 3000, res = 350);
par(mfrow=c(2,1))
par(mar = c(0, 4, 2, 0))
for (set in 1:nSets)
{
  plot(sampleTrees[[set]], main = paste("Sample clustering on all genes in", setLabels[set]),
       xlab="", sub="", cex = 0.7);
  abline(h=cutHeights[set], col = "red");
}
dev.off();


# for (set in 1:nSets)
# {
#   # Find clusters cut by the line
#   labels = cutreeStatic(sampleTrees[[1]], cutHeight = cutHeights)
#   # Keep the largest one (labeled by the number 1)
#   keep = (labels==1)
#   multiExpr[[set]]$data = multiExpr[[set]]$data[keep, ]
# }
# collectGarbage();
# # Check the size of the leftover data
# exprSize = checkSets(multiExpr)
# exprSize  
# This part doesnt work well for some reason, manually reove samples taht were outliers

multiExpr[[1]]$data <- multiExpr[[1]]$data[-which(row.names(multiExpr[[1]]$data)=="A5748"),]
multiExpr[[2]]$data <- multiExpr[[2]]$data[-which(row.names(multiExpr[[2]]$data)=="A5141"),]
exprSize = checkSets(multiExpr)
exprSize 
```


```{r}

# Form a multi-set structure that will hold the clinical traits.
Traits = vector(mode="list", length = nSets);
for (set in 1:nSets)
{
  setSamples = rownames(multiExpr[[set]]$data);
  traitRows = match(setSamples, rownames(trait1));
  Traits[[set]] = list(data = trait1[traitRows, ]);
  rownames(Traits[[set]]$data) = row.names(trait1[traitRows, ]);
}
collectGarbage();
# Define data set dimensions
nGenes = exprSize$nGenes;
nSamples = exprSize$nSamples;
save(multiExpr, Traits, nGenes, nSamples, setLabels, exprSize,  file = "Consensus-dataInput.RData");

```

#Data Simulation
```{r}
# number of samples in the training data
no.obs=85

```


```{r}
GS1= as.numeric(cor(trait1$Group, t(dat1)))
p.Standard=corPvalueFisher(GS1, nSamples =length(trait1$Group) )
p.Standard2=p.Standard
p.Standard2[is.na(p.Standard)]=1
q.Standard=qvalue(p.Standard2)$qvalues
table(q.Standard<.8)

# here we define the adjacency matrix using soft thresholding with beta=6
ADJ1=abs(cor(dat1,use="p"))^6
# When you have relatively few genes (<5000) use the following code
k=as.vector(apply(ADJ1,2,sum, na.rm=T))

# Plot a histogram of k and a scale free topology plot
sizeGrWindow(10,5)
par(mfrow=c(1,2))
hist(k)
scaleFreePlot(k, main="Check scale free topology\n")
dat1=dat1[, rank(-k,ties.method="first" )<=3600]
# Turn adjacency into a measure of dissimilarity
dissADJ=1-ADJ1
dissTOM=TOMdist(ADJ1)
collectGarbage()

```

```{r}
pam3=pam(as.dist(dissADJ), 3)
table(pam3$clustering)
# 1  2  3 
# 60 18  7 
pam4=pam(as.dist(dissADJ), 4)
table(pam4$clustering)
#  1  2  3  4 
# 49 17 12  7 
pam5=pam(as.dist(dissADJ), 5)
pam6=pam(as.dist(dissADJ), 6)
# # Cross-tabulte the detected and the true (simulated) module membership:
# table(pam4$clustering, truemodule)
# table(pam5$clustering, truemodule)
# table(pam6$clustering, truemodule)
pamTOM3=pam(as.dist(dissTOM), 4)
table(pamTOM3$clustering)
# 1  2  3  4
# 64 13  1  7 
pamTOM4=pam(as.dist(dissTOM), 4)
table(pamTOM4$clustering)
#  1  2  3  4 
# 26 33 13 13 
pamTOM5=pam(as.dist(dissTOM), 5)
pamTOM6=pam(as.dist(dissTOM), 6)
# # Cross-tabulte the detected and the true (simulated) module membership:
# table(pamTOM4$clustering, truemodule)
# table(pamTOM5$clustering, truemodule)
# table(pamTOM6$clustering, truemodule)
```


```{r}
# Choose a set of soft-thresholding powers
powers = c(seq(4,10,by=1), seq(12,20, by=2));
# Initialize a list to hold the results of scale-free analysis
powerTables = vector(mode = "list", length = nSets);
# Call the network topology analysis function for each set in turn
for (set in 1:nSets)
powerTables[[set]] = list(data = pickSoftThreshold(multiExpr[[set]]$data, powerVector=powers,
verbose = 2)[[2]]);
collectGarbage();
# Plot the results:
colors = c("black", "red")
# Will plot these columns of the returned scale free analysis tables
plotCols = c(2,5,6,7)
colNames = c("Scale Free Topology Model Fit", "Mean connectivity", "Median connectivity",
"Max connectivity");
# Get the minima and maxima of the plotted points
ylim = matrix(NA, nrow = 2, ncol = 4);
for (set in 1:nSets)
{
for (col in 1:length(plotCols))
{
ylim[1, col] = min(ylim[1, col], powerTables[[set]]$data[, plotCols[col]], na.rm = TRUE);
ylim[2, col] = max(ylim[2, col], powerTables[[set]]$data[, plotCols[col]], na.rm = TRUE);
}
}
# Plot the quantities in the chosen columns vs. the soft thresholding power
png(file = "ScaleFree topology model fit.png", width = 2000, height = 3000, res = 350);
sizeGrWindow(8, 6)
#pdf(file = "Plots/scaleFreeAnalysis.pdf", wi = 8, he = 6)
par(mfcol = c(2,2));
par(mar = c(4.2, 4.2 , 2.2, 0.5))
cex1 = 0.7;
for (col in 1:length(plotCols)) for (set in 1:nSets)
{
if (set==1)
{
plot(powerTables[[set]]$data[,1], -sign(powerTables[[set]]$data[,3])*powerTables[[set]]$data[,2],
xlab="Soft Threshold (power)",ylab=colNames[col],type="n", ylim = ylim[, col],
main = colNames[col]);
addGrid();
}
if (col==1)
{
text(powerTables[[set]]$data[,1], -sign(powerTables[[set]]$data[,3])*powerTables[[set]]$data[,2],
labels=powers,cex=cex1,col=colors[set]);
} else
text(powerTables[[set]]$data[,1], powerTables[[set]]$data[,plotCols[col]],
labels=powers,cex=cex1,col=colors[set]);
if (col==1)
{
legend("bottomright", legend = setLabels, col = colors, pch = 20) ;
} else
legend("topright", legend = setLabels, col = colors, pch = 20) ;
}
dev.off();

```
# Calculation of network adjacencies
```{r}
########## We choose the power 6 for both sets.
softPower = 6
# Initialize an appropriate array to hold the adjacencies
adjacencies = array(0, dim = c(nSets, nGenes, nGenes));
# Calculate adjacencies in each individual data set
for (set in 1:nSets)
adjacencies[set, , ] = abs(cor(multiExpr[[set]]$data, use = "p"))^softPower;
```

# Calculation of Topological Overlap

```{r}
# Initialize an appropriate array to hold the TOMs
TOM = array(0, dim = c(nSets, nGenes, nGenes));
# Calculate TOMs in each individual data set
for (set in 1:nSets)
TOM[set, , ] = TOMsimilarity(adjacencies[set, , ]);
```
Scaling of Topological Overlap Matrices to make them comparable across sets
```{r}
# Define the reference percentile
scaleP = 0.95
# Set RNG seed for reproducibility of sampling
set.seed(12345)
# Sample sufficiently large number of TOM entries
nSamples = as.integer(1/(1-scaleP) * 1000);
# Choose the sampled TOM entries
scaleSample = sample(nGenes*(nGenes-1)/2, size = nSamples, replace = TRUE)
TOMScalingSamples = list();
# These are TOM values at reference percentile
scaleQuant = rep(1, nSets)
# Scaling powers to equalize reference TOM values
scalePowers = rep(1, nSets)
# Loop over sets
for (set in 1:nSets)
{
# Select the sampled TOM entries
TOMScalingSamples[[set]] = as.dist(TOM[set, , ])[scaleSample]
# Calculate the 95th percentile
scaleQuant[set] = quantile(TOMScalingSamples[[set]],
probs = scaleP, type = 8);
# Scale the male TOM
if (set>1)
{
scalePowers[set] = log(scaleQuant[1])/log(scaleQuant[set]);
TOM[set, ,] = TOM[set, ,]^scalePowers[set];
}
}
```

Q-Q plots for the scaled data
```{r}
# For plotting, also scale the sampled TOM entries
scaledTOMSamples = list();
for (set in 1:nSets)
scaledTOMSamples[[set]] = TOMScalingSamples[[set]]^scalePowers[set]
# Open a suitably sized graphics window
sizeGrWindow(6,6)
#pdf(file = "Plots/TOMScaling-QQPlot.pdf", wi = 6, he = 6);
# qq plot of the unscaled samples
qqUnscaled = qqplot(TOMScalingSamples[[1]], TOMScalingSamples[[2]], plot.it = TRUE, cex = 0.6,
xlab = paste("TOM in", setLabels[1]), ylab = paste("TOM in", setLabels[2]),
main = "Q-Q plot of TOM", pch = 20)
# qq plot of the scaled samples
qqScaled = qqplot(scaledTOMSamples[[1]], scaledTOMSamples[[2]], plot.it = FALSE)
points(qqScaled$x, qqScaled$y, col = "red", cex = 0.6, pch = 20);
abline(a=0, b=1, col = "blue")
legend("topleft", legend = c("Unscaled TOM", "Scaled TOM"), pch = 20, col = c("black", "red"))
dev.off()
```

```{r}
consensusTOM = pmin(TOM[1, , ], TOM[2, , ])

```

```{r}
# Clustering  
consTree = hclust(as.dist(1-consensusTOM), method = "average");
# We like large modules, so we set the minimum module size relatively high:
minModuleSize = 3;
# Module identification using dynamic tree cut:
unmergedLabels = cutreeDynamic(dendro = consTree, distM = 1-consensusTOM,
deepSplit = 2, cutHeight = 0.99,
minClusterSize = minModuleSize,
pamRespectsDendro = FALSE );
unmergedColors = labels2colors(unmergedLabels)

```

TOM based clustering is giving very few clusters (one) with minimum size of 3. therefore use heirarchical


```{r}
hierADJ=hclust(as.dist(dissADJ), method="average" )
# Plot the resulting clustering tree together with the true color assignment
sizeGrWindow(10,5);
plotDendroAndColors(hierADJ, colors = data.frame(dat1), dendroLabels = FALSE, hang = 0.03, main = "Gene hierarchical clustering dendrogram and simulated module colors" )

```



```{r}
net = blockwiseConsensusModules(
multiExpr, power = 2, minModuleSize = 4, deepSplit = 2,
pamRespectsDendro = FALSE,
mergeCutHeight = 0.25, numericLabels = TRUE,
minKMEtoStay = 0,
saveTOMs = TRUE, verbose = 5)
```

```{r}
names(net)
# [1] "colors"                    "unmergedColors"           
#  [3] "multiMEs"                  "goodSamples"              
#  [5] "goodGenes"                 "dendrograms"              
#  [7] "TOMFiles"                  "blockGenes"               
#  [9] "blocks"                    "originCount"              
# [11] "networkCalibrationSamples" "individualTOMInfo"        
# [13] "consensusTOMInfo"          "consensusQuantile" 
```

```{r}
consMEs = net$multiMEs;
moduleLabels = net$colors;
# Convert the numeric labels to color labels
moduleColors = labels2colors(moduleLabels)
consTree = net$dendrograms[[1]];
```

```{r}

png(file = "Plots/ConsensusDendrogram-auto.pdf", wi = 8, he = 6)
plotDendroAndColors(consTree, moduleColors,
"Module colors",
dendroLabels = FALSE, #hang = 0.03,
addGuide = TRUE, #guideHang = 0.05,
main = "Consensus gene dendrogram and module colors")
dev.off()

```

```{r}
# Calculate module eigengenes
unmergedMEs = multiSetMEs(multiExpr, colors = NULL, universalColors = net$colors)
# Calculate consensus dissimilarity of consensus module eigengenes
consMEDiss = consensusMEDissimilarity(unmergedMEs);
# Cluster consensus modules
consMETree = hclust(as.dist(consMEDiss), method = "average");
# Plot the result
png("colorsDendro.png")
sizeGrWindow(7,6)
par(mfrow = c(1,1))
plot(consMETree, main = "Consensus clustering of consensus module eigengenes",
xlab = "", sub = "")
abline(h=0.94, col = "red")
dev.off()
```

```{r}
lnames = save(GS1, consensusTOM, dissADJ, multiExpr, net, unmergedMEs, file = "Consensus-NetworkConstruction-auto.RData")
```

Relating consensus modules to IBS set-speci c modules
```{r}

```

